### PR TITLE
Feature/cozy stack backend

### PR DESCRIPTION
**Pull request by thoumasse**
## Purpose

Implement authentication against CozyStack and backends to read from and write to CozyStack database.

### CODE REVIEW

File: src/ralph/backends/data/cozystack.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                response = client.post(url, json=kwargs)
+                response.raise_for_status()
+
+                json_response = response.json()
+
+                documents = json_response["docs"]
+
+                # query object is used to pass next and bookmark
+                query.next = json_response["next"]
+                query.bookmark = json_response["bookmark"]
```
Comment (pierrocknroll) [inline src/ralph/backends/data/cozystack.py:194]: Il faudra gérer les cas d'erreurs où ces valeurs n'existent pas dans json_response.

File: src/ralph/backends/data/cozystack.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+class CozyStackDataBackend(
+    BaseDataBackend[Settings, CozyStackQuery], Writable, Listable
+):
+    """CozyStack data backend."""
+
+    name = "cozy-stack"
+    unsupported_operation_types = {BaseOperationType.APPEND}
+
+    def __init__(self, settings: Optional[Settings] = None):
```
Comment (pierrocknroll) [inline src/ralph/backends/data/cozystack.py:118]: C'est un choix d'architecture mais ta classe CozyStackDataBackend est très couplée à CozyStackClient, peut-être faudrait-il abstraire un peu (pour les tests notamment), que CozyStackClient soit une variable de l'objet CozyStackDataBackend et que CozyStackClient ait des méthodes du genre get_all_doctypes et put_documents. Mais ton idée d'utiliser with est sympa aussi.

File: src/ralph/backends/data/cozystack.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+            elif operation_type == BaseOperationType.UPDATE:
+                preparedItem["_rev"] = ""
+                preparedItem["source"] = item
+
+            elif operation_type == BaseOperationType.DELETE:
+                preparedItem["_rev"] = ""
+                preparedItem["_deleted"] = True
+
+            preparedData.append(preparedItem)
```
Comment (pierrocknroll) [inline src/ralph/backends/data/cozystack.py:228]: Peut-être extraire cela dans une méthode séparée.

File: src/ralph/models/cozy.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    def _check_bearer_token(cls, v: str) -> str:
+        if "Bearer" not in v:
+            raise ValueError("must contain 'Bearer'")
+
+        return v
+
+    @field_validator('cookie')
+    @classmethod
+    def _check_cozysessid_cookie(cls, v: str) -> str:
+        if "cozysessid" not in v:
```
Comment (pierrocknroll) [inline src/ralph/models/cozy.py:30]: Peut-être faudrait-il mettre cozysessid dans une constante quelque part

**Review by pierrocknroll — COMMENTED** [review]
Bien joué, c'est bien organisé et tu as su réutiliser ce qui est prévu dans Ralph sans trop le modifier, c'est cool. Je t'ai mis quelques retours de pinaillages

File: src/ralph/backends/data/cozystack.py (hunk 2 above)
Comment (thoumasse) [reply src/ralph/backends/data/cozystack.py:118]: Yes je pensais faire une classe pour encapsuler toutes ces opérations dans des méthodes avec un plus haut niveau d'abstraction type `get_all_doctypes` etc. J'en profiterai pour en faire une variable comme dans les autres backends

File: src/ralph/backends/data/cozystack.py (hunk 1 above)
Comment (thoumasse) [reply src/ralph/backends/data/cozystack.py:194]: Il me semble que toutes les réponses doivent contenir ces champs mais ça ne coûte rien de les vérifier

File: src/ralph/api/auth/cozy.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
+"""OpenID Connect authentication tool for the Ralph API."""
```
[the remark this answers is not in the document]
Comment (thoumasse) [reply src/ralph/api/auth/cozy.py:1]: Bien vu

File: src/ralph/backends/data/cozystack.py (hunk 1 above)
Comment (pierrocknroll) [reply src/ralph/backends/data/cozystack.py:194]: J'ai tendance à faire du Defense Programming, je crois fort en la loi de Murphy :D

File: src/ralph/backends/lrs/cozystack.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        prefix = f"source.{target_field}"
+
+        if agent_params.get("mbox"):
+            query_filters[f"{prefix}.mbox"] = agent_params["mbox"]
+
+        if agent_params.get("mbox_sha1sum"):
+            query_filters[f"{prefix}.mbox_sha1sum"] = agent_params["mbox_sha1sum"]
+
+        if agent_params.get("openid"):
+            query_filters[f"{prefix}.openid"] = agent_params["openid"]
```
Comment (pierrocknroll) [inline src/ralph/backends/lrs/cozystack.py:74]: Tu peux simplifier cela par une boucle

File: src/ralph/backends/lrs/cozystack.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+            )
+
+        except (BackendException, BackendParameterException) as error:
+            logger.error("Failed to read from CozyStack")
+            raise error
+
+        search_after = query.bookmark if query.next else None
+
+        return StatementQueryResult(
+            statements=[document["source"] for document in response],
```
Comment (pierrocknroll) [inline src/ralph/backends/lrs/cozystack.py:149]: Sûr qu'il y aura toujours un "source" dans chaque document ?

File: tests/api/cozystack/test_statements_post.py [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+    }
+
+
+@pytest.mark.anyio
+async def test_api_statements_post_single_statement_directly(
+    client, init_cozystack_db_and_monkeypatch_backend, cozy_auth_token
+):
+    """Test the post statements API route with one statement."""
+    init_cozystack_db_and_monkeypatch_backend()
+    statement = mock_statement()
```
Comment (pierrocknroll) [inline tests/api/cozystack/test_statements_post.py:47]: Il faudrait tester le passage de statements pêtés

File: tests/api/cozystack/test_statements_post.py [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+        headers={"X-Auth-Token": f"Bearer {cozy_auth_token}"},
+        json=statement,
+    )
+
+    assert response.status_code == 200
+
+    response = await client.get(
+        "/xAPI/statements/",
+        headers={"X-Auth-Token": f"Bearer {cozy_auth_token}"},
+    )
```
Comment (pierrocknroll) [inline tests/api/cozystack/test_statements_post.py:98]: Tu ne peux récupérer le statement juste créé que via un GET global ?

File: tests/backends/cozystack/test_client.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+async def test_cozystack_client_list_all_doctypes(cozystack_custom, cozy_auth_target):
+    cozystack_custom()
+    client = CozyStackClient(COZYSTACK_TEST_DOCTYPE)
+
+    response = client.list_all_doctypes(target=cozy_auth_target)
+    assert "io.cozy.learningrecords" in response
+
+
+@pytest.mark.anyio
+async def test_cozystack_client_find(
```
Comment (pierrocknroll) [inline tests/backends/cozystack/test_client.py:54]: Je ne vais pas te la refaire à chaque fois mais il me semble qu'ici il y a plein de tests potentiels à ajouter sur des valeurs incorrectes.

File: tests/backends/cozystack/test_client.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    cozystack_custom()
+    client = CozyStackClient(COZYSTACK_TEST_DOCTYPE)
+
+    # index statements
+    response = client.bulk_operation(
+        target=cozy_auth_target, data=statements, operation_type=BaseOperationType.INDEX
+    )
+    assert response == len(statements)
+
+    # try ton index twice
```
Comment (pierrocknroll) [inline tests/backends/cozystack/test_client.py:129]: to

**Review by pierrocknroll — COMMENTED** [review]
Coucou @thoumasse 
Ca y'est j'ai regardé la partie tests de ta PR ! Bon, je n'ai pas regardé ligne par ligne car c'est énorme et je te fais confiance, ça a du être long et pénible à faire, bravo ! Tu as couvert une bonne partie des cas, je pense que tu testes toujours les cas optimistes, comme tu le verras dans mes commentaires je pense qu'il manque des cas d'erreur.
Aussi, je pense que globalement il faudrait ajouter du typage dans les signatures de tes méthodes de tests, cela serait + propre et aiderait aussi à la compréhension de ceux-ci. Certains n'ont pas non plus de doc associée. C'est pas très grave car ils sont compréhensibles pour la plupart.
Encore bravo ^^ On peut passer à la suite maintenant, mais n'oublie pas aussi les quelques commentaires restant de mes précédentes reviews.

File: tests/api/cozystack/test_statements_get.py [hunk 12]
```diff
@@ -0,0 +1,642 @@
...
+    # normal case
+    response = await client.get(
+        "/xAPI/statements/?ascending=true",
+        headers={"X-Auth-Token": f"Bearer {cozy_auth_token}"},
+    )
+
+    assert response.status_code == 200
+    assert response.json() == {"statements": [statements[0], statements[1]]}
+
+    # edge case: wrong ascending param
```
Comment (pierrocknroll) [inline tests/api/cozystack/test_statements_get.py:83]: Je vais être chiant, mais il est généralement conseillé de séparer les tests dans des méthodes dédiées (pour des questions de lisibilité, éviter les interférences, + facile à débugguer).

**Comment by pierrocknroll** [comment]
Coucou @thoumasse 
J'ai regardé tes récents changements suite à mes commentaires pour les tests, et c'est top d'avoir ajouté les edge cases et le typage ! Je vois que tu as dû adapter le code pour faire passer les edge cases, c'est plutôt bon signe que les valeurs exotiques doivent être exclues.
Et sinon, je ne connaissais pas du tout ce concept de TypeAdapter et c'est une bonne idée de l'avoir mis, bien joué, je retiens ce truc pour mes projets ^^ C'est je pense une bonne chose quand on reçoit des données de l'extérieur et qu'un Model n'est pas nécessaire, mais ça n'empêche pas qu'il faut typer ses variables ^^

File: src/ralph/api/auth/cozy.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        iss (str): Issuer Identifier for the Issuer of the response.
+        sub (str): Subject Identifier.
+        aud (str): Audience(s) that this ID Token is intended for.
+        iat (int): Time at which the JWT was issued.
+        scope (str): Scope(s) for resource authorization.
+        target (str): Target for storing the statements.
+        session_id (int): TODO
+    """
+
+    aud: List[str]
```
[the remark this answers is not in the document]
Comment (pierrocknroll) [reply src/ralph/api/auth/cozy.py:46]: Je pense que c'est l'objet d'une autre PR donc je ferme ce commentaire :)

File: src/ralph/api/auth/cozy.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+    Return:
+        Dict: Decoded token.
+
+    Raises:
+        HTTPException: When decoding fails.
+    """
+    try:
+        if not isinstance(x_auth_token, str):
+            raise ValueError("X-Auth-Token must be a string")
```
[the remark this answers is not in the document]
Comment (pierrocknroll) [reply src/ralph/api/auth/cozy.py:64]: Tu as raison, on peut laisser :)

File: src/ralph/backends/cozystack/client.py [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+import httpx
+from fastapi import HTTPException, status
+from pydantic import ValidationError
+
+from ralph.backends.cozystack import exceptions
+from ralph.backends.data.base import BaseOperationType
+from ralph.models.cozy import CozyAuthData
+
+
+class CozyStackHttpClient:
```
[the remark this answers is not in the document]
Comment (pierrocknroll) [reply src/ralph/backends/cozystack/client.py:25]: Cool !

File: tests/api/cozystack/test_statements_post.py (hunk 9 above)
Comment (pierrocknroll) [reply tests/api/cozystack/test_statements_post.py:119]: Je me posais simplement la question dans le cadre des tests s'il est + pertinent de vérifier si la donnée est présente en appelant le endpoint de liste globalement, ou en précisant l'ID du statement fraîchement créé (ou les 2, quelque part)

Comment (thoumasse) [reply tests/api/cozystack/test_statements_post.py:119]: C'est fait

File: tests/api/cozystack/test_statements_get.py (hunk 12 above)
Comment (thoumasse) [reply tests/api/cozystack/test_statements_get.py]: C'est fait !
